backend/tests/test_user_class.py

```diff
@@ -0,0 +1,79 @@
+import json
+import os
+
+import pytest
+from hypothesis import assume, given, settings
+from hypothesis.strategies import emails, integers, none, one_of, text
+
+from ..exceptions import UserNameTakenError
+from ..structs import UserInfo
+from .app_fixtures import *
+from .user_fixtures import *
+
+
+@given(pk=integers(), display_name=text(min_size=1), email=emails(), password=one_of(text(min_size=1), none()))
+@settings(max_examples=10)  # something about this takes forever!
+@pytest.mark.filterwarnings(
+    'ignore:.*'
+)  # this test warns that the fixture is not reset each test run and recommends using a context manager, which we do. so i want to disable just that warning, but the regex filtering any more specific than that isnt working TODO
+def test_serialize_deserialize_roundtrip_through_redis(redis, pk, display_name, email, password):
+    user = UserInfo(pk=pk, display_name=display_name, email=email)
+    user.set_password(password)
+    with hypothesis_safe_redis(redis) as redis:
+        redis.hmset('key', user.serialize())
+        round_tripped_user = UserInfo.deserialize(redis.hgetall('key'))
+        assert user == round_tripped_user
+
+
+@given(pk=integers(), display_name=text(min_size=1), email=emails(), password=one_of(text(min_size=1), none()))
+@settings(max_examples=10)  # something about this takes forever!
+def test_deserialize_not_from_redis(pk, display_name, email, password):
+    u = UserInfo(pk=pk, display_name=display_name, email=email)
+    u.set_password(password)
+    ser = u.serialize(serialize_method=json.dumps)
+    u2 = UserInfo.deserialize(json.loads(ser), from_redis=False)
+    assert u == u2
+
+
+@given(email=emails())
+def test_is_anonymous(email):
+    u = UserInfo(pk=1, email=email)
+    assert u.is_anonymous() == (email == '')
+
+
+@given(one_of(none(), text(min_size=1)), text(min_size=1))
+def test_password_checking_works(password, wrong_password):
+    assume(password != wrong_password)
+    u = UserInfo(pk=1)
+
+    u.set_password(password)
+    assert not u.check_password(wrong_password)
+    if password is None:
+        assert not u.check_password(password)
+    else:
+        assert u.check_password(password)
+
+
+if os.environ.get('TRAVIS'):
+    # test password hashing times out in CI, little too long
+    # so just cut it short for now
+    # TODO investigate using a CI profile instead of this manual call
+    reduced_max = settings(max_examples=1)
+else:
+    reduced_max = settings(max_examples=10)
+test_password_checking_works = reduced_max(test_password_checking_works)
+
+
+def test_update_user_happy_path(user, redis):
+    u = user()
+    u.email = 'newemail@newtown.com'
+    u.update_user(redis)
+    assert u.email == 'newemail@newtown.com'
+
+
+def test_update_user_fails_if_overwriting(user, redis):
+    u1 = user()
+    u2 = user()
+    u2.pk = u1.pk
+    with pytest.raises(UserNameTakenError):
+        u2.update_user(redis)
```

Comment:
Confirming my understanding: This attempts to the set the `pk` of u2 to that same as u1. And this results in a User_Name_TakenError? why? isn't the pk taken? is the difference not worth having a different exception? or is this just standard practice and I just don't know what i'm talking about?

Reply:
Mm yea this is a bit of an artifact, is now confusing.

The thought here is

I was worried about someone getting a user object and altering its pk (since its python and nothing would stop you from just doing that) and then saving it, overwriting the information (normal and you didn't need me to explain that that would be bad).  In a regular DB that would be that.

But the way that I check to see if the PK is in use is, on save, I look up in the EmailByPk table and see if the email you say you are matches the PK you claim to have (and its called username because email used to be called username).  Why check that instead of checking the UsersByUserPK table?  Because that one didn't exist when I wrote this code :).
So in that order of operations, it was that the username was already claimed by another pk - hence the exception name

Now that I've cleaned all this up, yes certainly this should be PKAlreadyTaken, and I should look up in the PK table to see if your email matches.
Except what if you are changing your email?  Maybe the right thing to do now is check that the pk is equal to current_user.pk?  Except then you cant call save from scripts.  Well maybe its ok that PK _and_ email are both immutable - if you want a new email address you make a new account and live your life?  What would happen in ciqapp?  Ah, both pk and email must be globally unique.  
So if I took pk1==Jacob@Jacob.com, set his pk to 2, that would fail because then id have pks 1 _and_ 2 with email Jacob@Jacob.com
If I took him and set his pk to 99999999, one that does exist, then ok, except I'd break the PkByEmail mapping - ok maybe i just rectify this on save?
If I took him and changed his email and his pk, then I might overwrite someone, not sure how to solve

What do you think?